external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CachedKafkaProducerSuite.scala

```diff
@@ -18,60 +18,206 @@
+  test("Should return the new instance on calling acquire with different params.") {
+    val kafkaParams: ju.HashMap[String, Object] = generateKafkaParams
+    val producer = CachedKafkaProducer.acquire(kafkaParams)
+    kafkaParams.remove("ack") // mutate the kafka params.
+    val producer2 = CachedKafkaProducer.acquire(kafkaParams)
+    assert(producer.kafkaProducer != producer2.kafkaProducer)
+    assert(producer.getInUseCount == 1)
+    assert(producer2.getInUseCount == 1)
+    val map = CachedKafkaProducer.getAsMap
+    assert(map.size == 2)
+  }
+
+  test("Automatically remove a failing kafka producer from cache.") {
+    import testImplicits._
+    val df = Seq[(String, String)](null.asInstanceOf[String] -> "1").toDF("topic", "value")
+    val ex = intercept[SparkException] {
+      // This will fail because the service is not reachable.
+      df.write
+        .format("kafka")
+        .option("topic", "topic")
+        .option("kafka.retries", "1")
+        .option("kafka.max.block.ms", "2")
+        .option("kafka.bootstrap.servers", "12.0.0.1:39022")
+        .save()
+    }
+    assert(ex.getMessage.contains("org.apache.kafka.common.errors.TimeoutException"),
+      "Spark command should fail due to service not reachable.")
+    // Since failing kafka producer is released on error and also invalidated, it should not be in
+    // cache.
+    val map = CachedKafkaProducer.getAsMap
+    assert(map.size == 0)
+  }
+
+  test("Should not close a producer in-use.") {
+    val kafkaParams: ju.HashMap[String, Object] = generateKafkaParams
+    val producer: CachedKafkaProducer = CachedKafkaProducer.acquire(kafkaParams)
+    producer.kafkaProducer // initializing the producer.
+    assert(producer.getInUseCount == 1)
+    // Explicitly cause the producer from guava cache to be evicted.
+    CachedKafkaProducer.evict(producer.getKafkaParams)
+    assert(producer.getInUseCount == 1)
+    assert(!producer.isClosed, "An in-use producer should not be closed.")
+  }
+
+  private def generateKafkaParams: ju.HashMap[String, Object] = {
     val kafkaParams = new ju.HashMap[String, Object]()
-    kafkaParams.put("acks", "0")
+    kafkaParams.put("ack", "0")
     kafkaParams.put("bootstrap.servers", "127.0.0.1:9022")
     kafkaParams.put("key.serializer", classOf[ByteArraySerializer].getName)
     kafkaParams.put("value.serializer", classOf[ByteArraySerializer].getName)
-    val producer: KP = CachedKafkaProducer.getOrCreate(kafkaParams)
-    kafkaParams.put("acks", "1")
-    val producer2: KP = CachedKafkaProducer.getOrCreate(kafkaParams)
-    // With updated conf, a new producer instance should be created.
-    assert(producer != producer2)
-
-    val cacheMap = PrivateMethod[ConcurrentMap[Seq[(String, Object)], KP]]('getAsMap)
-    val map = CachedKafkaProducer.invokePrivate(cacheMap())
-    assert(map.size == 2)
+    kafkaParams
+  }
+}
+
+class CachedKafkaProducerStressSuite extends KafkaContinuousTest with KafkaTest {
```

Comment:
Like the direction but maybe some fine tuning would be good because I receive this exception on my local machine:
```
Caused by: org.apache.kafka.common.KafkaException: java.io.IOException: Too many open files
```
On a beefy executor maybe 20 different TopicPartition is realistic. I think the number of usages can be set high just like in the consumer side test.

Reply:
Not sure whether its forgotten or under implementation but this hasn't been resolved and I still have the same issue.

Reply:
I was not able to see that error, on my laptop(which has very low spec). My guess is, it happens due to too many producer object creation. I can reduce it, in the test.
But maybe you can increase the ulimit on your system, the default is a bit low.

A side note, the test does not fail, even without synchronisation blocks, on my system. Am I missing something?

Reply:
I have a MacOS Mojave with default settings. I've not done any deadlock or other magic when this issue came. We've tried it on 2 other machines and produced the same exception.

> But maybe you can increase the ulimit on your system, the default is a bit low.

Expecting maybe several thousand contributors to adapt the system to a single test is not a really classical approach. Will check the the updated test...

Reply:
Interestingly, I too have a MacOS Mojave, and never saw that error. BTW, are you still getting it with the updated test? In that case, I can lower the number of producers further.

Reply:
Re-tested `test("concurrent use of CachedKafkaProducer")` and after ~10239 open file descriptors the same exception came. In `test("SPARK-23623: concurrent use of KafkaDataConsumer")` this number is fluctuating around 500. You can check it by adding the following code after `aquire`:
```
import java.lang.management.ManagementFactory
import com.sun.management.UnixOperatingSystemMXBean
System.out.println("Number of open fd: " + ManagementFactory.getOperatingSystemMXBean.asInstanceOf[UnixOperatingSystemMXBean].getOpenFileDescriptorCount)
```

Reply:
I've done the following to fix this:
```
    val data = (1 to 100).map(_.toString)
...
      val futuresAcquire = (1 to 10 * numConcurrentProducers).map { i =>
        threadPool.submit(new Runnable {
          override def run(): Unit = {
            val producer = acquire(i % numConcurrentProducers + 1)
            data.foreach { d =>
              val record = new ProducerRecord[Array[Byte], Array[Byte]](topic, 0, null, d.getBytes)
              producer.kafkaProducer.send(record)
            }
            release(producer)
          }
        })
      }
```
`toBeReleasedQueue` removed completely.
This way open files is fluctuating around 500-600 and more real life scenario.